File: 8_graph_shortest_path/이준우/가장 먼 노드.py
import heapq

INF = int(1e9)


def solution(n, edge):
    graph = [[] for _ in range(n + 1)]
    distance = [INF] * (n + 1)

    for a, b in edge:
        graph[a].append(b)
        graph[b].append(a)

    pq = []
    heapq.heappush(pq, (0, 1))
    distance[1] = 0

    while pq:
        dist, now = heapq.heappop(pq)
        if distance[now] < dist:
            print(dist, distance[now])
            continue

        for next in graph[now]:
            cost = distance[now] + 1
            if distance[next] > cost:
                distance[next] = cost
                heapq.heappush(pq, (cost, next))

    max_value = max(list(filter(lambda x: x != INF, distance)))
    return distance.count(max_value)

File: 8_graph_shortest_path/이준우/가장 먼 노드.kt
class Solution {
    fun solution(n: Int, edge: Array<IntArray>): Int {
        val distance = IntArray(n + 1) { -1 }
        val graph = Array(n + 1) { mutableListOf<Int>() }
        for (e in edge) {
            val (a, b) = e
            graph[a].add(b)
            graph[b].add(a)
        }

        val q = ArrayDeque<Int>()
        q.add(1)
        distance[1] = 0

        while (q.isNotEmpty()) {
            val cur = q.removeFirst()

            for (next in graph[cur]) {
                if (distance[next] == -1) {
                    distance[next] = distance[cur] + 1
                    q.add(next)
                }
            }
        }

        val maxValue = distance.maxOrNull() ?: 0
        return distance.count { it == maxValue }
    }
}

/**
 * 가장 멀리 떨어진 노드의 갯수
 * 
 * BFS -> 처음에 목적지에 도달할 경우 그것이 최단거리
 */